nomad/structs/funcs.go

```diff
@@ -163,6 +166,18 @@ func AllocsFit(node *Node, allocs []*Allocation, netIdx *NetworkIndex, checkDevi
 				reservedCores[core] = struct{}{}
 			}
 		}
+
+		// Job will be nil in the scheduler, where we're not performing this check anyways
+		if checkDevices && alloc.Job != nil {
+			group := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+			for _, volReq := range group.Volumes {
+				hostVolumeClaims[volReq.Source]++
+				if volReq.AccessMode ==
+					CSIVolumeAccessMode(HostVolumeAccessModeSingleNodeSingleWriter) {
```

Comment:
I got a mental hiccup seeing `CSIVolume...(HostVolume...)` -- I'd think it should be

```suggestion
					HostVolumeAccessMode(HostVolumeAccessModeSingleNodeSingleWriter) {
```

Reply:
Unfortunately not (ref [`volumes.go#L100-L101`](https://github.com/hashicorp/nomad/blob/v1.9.4/nomad/structs/volumes.go#L100-L101)) because the type for `VolumeRequest` is named for the CSI mode.

```go
// VolumeRequest is a representation of a storage volume that a TaskGroup wishes
// to use.
type VolumeRequest struct {
	Name           string
	Type           string
	Source         string
	ReadOnly       bool
	Sticky         bool
	AccessMode     CSIVolumeAccessMode
	AttachmentMode CSIVolumeAttachmentMode
	MountOptions   *CSIMountOptions
	PerAlloc       bool
}
```

But both are "defined types" on string, which msgpack serializes to the same thing on the wire. So we _could_ swap out for a new shared `VolumeAccessMode` / `VolumeAttachmentMode` without breaking backcompat. Especially because in this case we didn't define types in the `api` package version of these fields. They're just strings there!

I'll do that, but under a separate PR.

Reply:
https://github.com/hashicorp/nomad/pull/24896